File: leetcode/python/leetcode_238_productofarrayexceptself.py
''' LeetCode - 238 - Product of Array Except Self
    Given an array of integers nums, return an array res such that
    res[i] is equal to the product of all the elements of nums except nums[i]
    Do not use division operator
    ex: nums = [1,2,3,4] => [24,12,8,6]

    Author: Thiru

    Time complexity: O(n)
    Space complexity: O(n) '''

''' Function to return the product of array except self
    Use identify of multiplication 1, to multiply the array from left to right
    Use identify of multiplication 1, to multiply the array from right to left
    Multiply the results from both of these arrays '''

from typing import List

def prodcutOfArrayExceptSelf(nums: List[int]) -> List[int]:
    size = len(nums)
    res = [1] * size

    for i in range(1, size):
        res[i] = res[i-1] * nums[i-1]

    r = 1
    for i in range(size-1, -1, -1):
        res[i] *= r
        r *= nums[i]

    return res

# main
nums = [1,2,3,4]
res = prodcutOfArrayExceptSelf(nums)
print(res)

File: leetcode75/cpp/leetcode238productofarrayexceptself.cpp
// LeetCode - 238 - Product of Array Except Self
// Given an array of integers nums, return an array res such that
// res[i] is equal to the product of all the elements of nums except nums[i]
// Do not use division operator
// ex: nums = [1,2,3,4] => [24,12,8,6]
//
// Author: Thiru
//
// Time complexity: O(n)
// Space complexity: O(n)

#include <iostream>
#include <vector>

// Function to return the product of array except self
// Use identify of multiplication 1, to multiply the array from left to right
// Use identify of multiplication 1, to multiply the array from right to left
// Multiply the results from both of these arrays

std::vector<int> prodcutOfArrayExceptSelf(const std::vector<int>& nums) {
  int size = nums.size();
  std::vector<int> res(size, 1);

  for (int i=1; i<size; i++) {
    res[i] = res[i-1] * nums[i-1];
  }

  int r = 1;
  for (int i=size-1; i>=0; i--) {
    res[i] = res[i] * r;
    r = r * nums[i];
  }

  return res;
}

// main
int main()
{
  std::vector<int> nums{1,2,3,4};
  std::vector<int> res = prodcutOfArrayExceptSelf(nums);

  for (const int& i: res) {
    std::cout<<i<<" ";
  }
  std::cout<<std::endl;

  return 0;
}
